File: card.py
import random
import math
import time
import os


def generate_heightmap(width, height):
	heightmap = [[0 for y in range(height)] for x in range(width)]

	MIN_HILL_RADIUS = 5
	MAX_HILL_RADIUS = 30

	for i in range(200):
		px = random.randint(0, width-1)     							# center of hill
		py = random.randint(0, height-1)    							# center of hill
		r =  random.randint(MIN_HILL_RADIUS, MAX_HILL_RADIUS)           # radius

		for x in range(width):
			for y in range(height):
				hill_height = r**2 - ((x-px)**2 + (y - py)**2)
				if hill_height > 0:
					new_height = heightmap[x][y] + hill_height
					heightmap[x][y] = new_height
	
	max_height = max([max(row) for row in heightmap])

	# flatten heightmap
	for x in range(width):
		for y in range(height):
			cur_height = heightmap[x][y]
			heightmap[x][y] = int((cur_height/max_height)*100)

	return heightmap


def generate_colormap(heightmap):
	width, height = len(heightmap), len(heightmap[0])

	max_height = max([max(row) for row in heightmap])

	colormap = [[0 for y in range(height)] for x in range(width)]
	for x in range(width):
		for y in range(height):
			grey_value = min(255, int((heightmap[x][y]/max_height)*255))
			rgb = (grey_value, grey_value, grey_value)
			colormap[x][y] = rgb

	for x in range(width):
		for y in range(height):

			rgb = colormap[x][y]
			rnd = random.uniform(-0.005, 0.005)

			# Coloring of terrain based on height
			if heightmap[x][y] <= max_height*(0.9+rnd):
				colormap[x][y] = (rgb[0]-40, rgb[1]-50, rgb[2]-40)
			if heightmap[x][y] <= max_height*(0.5+rnd):
				colormap[x][y] = (rgb[0]-10, rgb[1]+20, rgb[2]-10)
			if heightmap[x][y] <= max_height*(0.2+rnd):
				colormap[x][y] = (rgb[0]+120, rgb[1] + 80, rgb[2] + 40)
			if heightmap[x][y] <= max_height*0.1:
				colormap[x][y] = (rgb[0]-10, rgb[1]-10, rgb[2]+40)
			if heightmap[x][y] <= 5:
				 colormap[x][y] = (0, 0, 40)
			
			# Shading of terrain based on height of other blocks
			if (x > 0 and x < width and heightmap[x-1][y] > heightmap[x][y]):
				rgb = colormap[x][y]
				colormap[x][y] = (int(rgb[0]* 0.85),int(rgb[1]*0.85), int(rgb[2]*0.85))

	return colormap


def render(heightmap, colormap, p, height, horizon, scale_height, distance, fog_distance, screen_width, screen_height):
	# Default background color, doubles for sky
	buffer = [[(86+y-x, 213+y-x, 255+y-x) for y in range(screen_height)] for x in range(screen_width)]

	# Draw from back to the front
	for z in range(distance, 1, -1):
		# Find line on map. This calculation corresponds to a field of view of 90
		pleft  = (-z + p[0], -z + p[1])
		pright = (z + p[0], -z + p[1])
		
		# segment the line
		dx = (pright[0] - pleft[0]) / screen_width

		# Draw vertical lines
		for x in range(screen_width):
			height_on_screen = (height - heightmap[int(pleft[0])][int(pleft[1])]) / z * scale_height + horizon

			if (height_on_screen >= screen_height): continue
			if (height_on_screen < 0): height_on_screen = 0

			for y in range(math.floor(height_on_screen), math.floor(screen_height)):
				rgb = colormap[int(pleft[0])][int(pleft[1])]
				if z > fog_distance:
					m = 1-((fog_distance-z)/fog_distance)*1.5
					rgb = (int(rgb[0]*m),int(rgb[1]*m), int(rgb[2]*m))
				buffer[x][y] = rgb
			
			pleft = (pleft[0]+dx, pleft[1])
		
	return buffer


def draw(buffer, screenmap, screen_width, screen_height):
	line = "\033[?25l" + "\033[H"
	for y in range(0, screen_height, 2):
		for x in range(screen_width):
			char = "\033[38;2;{};{};{}m".format(*buffer[x][y])
			char += "\033[48;2;{};{};{}m▀".format(*buffer[x][y+1])
			line += char
		line += "\n"
	print(line[:-2], end='')


def fly():
	os.system("clear")
	screen_width =80
	screen_height = 48

	heightmap = generate_heightmap(200, 400)
	colormap = generate_colormap(heightmap)

	point = (100, 400)

	screenmap = [[None for y in range(screen_height)] for x in range(screen_width)]
	for i in range(int((400-100)-2)):
		buffer = render(heightmap, colormap, point, heightmap[point[0]-1][point[1]-5]+20, 7, 10, 100, 65, screen_width, screen_height)
		draw(buffer, screenmap, screen_width, screen_height)

		point = (point[0], point[1]-2)

fly()

File: card.cpp
#include <iostream>
#include <algorithm>
#include <cmath>
#include <time.h>
#include <unistd.h>


int* generate_heightmap(int width, int height) {
	int* heightmap = new int[width*height]{0};

	const int MIN_HILL_RADIUS = 5;
	const int MAX_HILL_RADIUS = 30;
	const int NR_OF_HILLS = 200;

	for (int i = 0; i < NR_OF_HILLS; i++) {
		int hill_x = rand() % width;
		int hill_y = rand() % height;
		int hill_radius = rand() % (MAX_HILL_RADIUS - MIN_HILL_RADIUS + 1) + MIN_HILL_RADIUS;

		for (int x = 0; x < width; x++) {
			for (int y = 0; y < height; y++) {
				double hill_height = pow(hill_radius, 2) - (pow(x - hill_x, 2) + pow(y - hill_y, 2));
				if (hill_height > 0) {
					double new_height = heightmap[y*width + x] + hill_height;
					heightmap[y*width + x] = int(new_height);
				}
			}
		}
	}

	double max_height = *std::max_element(heightmap, heightmap + width * height);

	for (int x = 0; x < width; x++) {
		for (int y = 0; y < height; y++) {
			int cur_height = heightmap[y*width + x];
			heightmap[y*width + x] = int((cur_height / max_height) * 100);
		}
	}

	return heightmap;
}


struct rgb {
	int r, g, b;
};

rgb* generate_colormap(int* heightmap, int map_width, int map_height) {
	double max_height = *std::max_element(heightmap, heightmap + map_width * map_height);

	rgb* colormap = new rgb[map_width*map_height];

	for (int x = 0; x < map_width; x++) {
		for (int y = 0; y < map_height; y++) {
			int grey_value = std::min(255, int((heightmap[y*map_width + x] / max_height) * 255));
			rgb color = { grey_value, grey_value, grey_value };
			colormap[y*map_width + x] = color;
		}
	}

	for (int x = 0; x < map_width; x++) {
		for (int y = 0; y < map_height; y++) {
			rgb color = colormap[y*map_width + x];
			int height = heightmap[y*map_width + x];

			if (height <= 5) {
				color = rgb{ 0, 0, 40 };
			} else if (height <= max_height * (0.1)) {
				color = rgb{ color.r - 10, color.g - 10, color.b + 40 };
			} else if (height <= max_height * (0.2)) {
				color = rgb{ color.r + 120, color.g + 80, color.b + 40 };
			} else if (height <= max_height * (0.5)) {
				color = rgb{ color.r - 10, color.g + 20, color.b - 10 };
			} else if (height <= max_height * (0.9)) {
				color = rgb{ color.r - 40, color.g - 50, color.b - 40 };
			}

			colormap[y*map_width + x] = color;

			if (x > 0 && x < map_width && heightmap[y*map_width + (x - 1)] > heightmap[y*map_width + x]) {
				rgb color = colormap[y*map_width + x];
				color.r *= 0.85;
				color.g *= 0.85;
				color.b *= 0.85;
                colormap[y*map_width + x] = color;
			}

		}
	}

	/**for (int x = 0; x < map_width; x++) {
		for (int y = 0; y < map_height; y++) {
			printf("%d %d %d | ", colormap[y*map_width + x].r, colormap[y*map_width + x].g, colormap[y*map_width + x].b);
		}
		printf("\n");
	}

	getchar();**/

	return colormap;

}

struct pos {
	double x, y;
};

rgb* render(int* heightmap, rgb* colormap, pos p, int height, int horizon, int scale_height, int distance, int fog_distance, int screen_width, int screen_height, int map_width, int map_height) {
	rgb* buffer = new rgb[screen_width*screen_height];

    int actual_height = height + heightmap[int(p.y)*map_width + int(p.x)];

	for (int x = 0; x < screen_width; x++) {
		for (int y = 0; y < screen_height; y++) {
			buffer[y*screen_width + x] = rgb{ 86 + y - x, 213 + y - x, 255 + y - x };
		}
	}

	for (int z = distance; z > 1; z--) {
		pos pleft = { -z + p.x, -z + p.y };
		pos pright = { z + p.x, -z + p.y };

		double dx = (pright.x - pleft.x) / screen_width;

		for (int x = 0; x < screen_width; x++) {
			double height_on_screen = (actual_height - heightmap[int(pleft.y)*map_width + int(pleft.x)]) / double(z) * scale_height + horizon;

			if (height_on_screen >= screen_height) { continue; }
			if (height_on_screen < 0) { height_on_screen = 0; }

			for (int y = floor(height_on_screen); y < floor(screen_height); y++) {
				rgb color = colormap[int(pleft.y)*map_width + int(pleft.x)];
				if (z > fog_distance) {
					double m = 1 - ((fog_distance - z) / double(fog_distance))*1.5;
					color = rgb{int(color.r*m), int(color.g*m), int(color.b*m) };
				}
				buffer[y*screen_width + x] = color;
			}
			
			pleft.x += dx;
		}
	}

	return buffer;
}

// Should be okay
void draw(rgb* buffer, int screen_width, int screen_height) {
	printf("\e[?25l\e[H");

	for (int y = 0; y < screen_height; y += 2) {
		for (int x = 0; x < screen_width; x++) {
			rgb color1 = buffer[y*screen_width + x];
			rgb color2 = buffer[(y + 1)*screen_width + x];
			printf("\e[38;2;%d;%d;%dm", color1.r, color1.g, color1.b);
			printf("\e[48;2;%d;%d;%dm▀", color2.r, color2.g, color2.b);
		}
	}
	fflush(stdout);
}

//Should also be okay
void fly() {
	system("clear");
    srand( time(0) );

	int screen_width = 80;
	int screen_height = 48;

	int map_width = 200;
	int map_height = 400;

	int* heightmap = generate_heightmap(map_width, map_height);
	rgb* colormap = generate_colormap(heightmap, map_width, map_height);

	pos point = { 100, 400 };

	rgb* screenmap = new rgb[screen_width*screen_height];

    for (; point.y > int(100-2); point.y-=2) {
        rgb* buffer = render(heightmap, colormap, point, 50, 7, 10, 100, 65, screen_width, screen_height, map_width, map_height);
        draw(buffer, screen_width, screen_height);
        usleep(50000);
	}
}

int main()
{
	fly();
	printf("\e[0m");
	return 0;
}
